src/Integration.Vsix/Suppression/LiveIssueFactory.cs

```diff
@@ -0,0 +1,143 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell.Interop;
+
+/* To map from a diagnostic to a SonarQube issue we need to work out the SQ moduleId
+ * corresponding to the MSBuild project.
+ *
+ * To speed things up, we cache the mapping from project file path to project id
+ * for the loaded projects (which means we need to rebuild the mapping when the
+ * solution changes).
+ *
+ * Note: this class does not monitor changes to the solution. If the solution/binding
+ * changes then create a new instance.
+ */
+
+namespace SonarLint.VisualStudio.Integration.Vsix.Suppression
+{
+    /// <summary>
+    /// Factory for <see cref="LiveIssue"/>s i.e. diagnostics that are decorated with
+    /// additional information required to map to issues in the SonarQube server format
+    /// </summary>
+    internal sealed class LiveIssueFactory
+    {
+        private readonly Workspace workspace;
+        private readonly IServiceProvider serviceProvider;
+
+        /// <summary>
+        /// Mapping from full project file path to the unique project id
+        /// </summary>
+        private Dictionary<string, string> projectPathToProjectIdMap;
+
+        public LiveIssueFactory(IServiceProvider serviceProvider)
+        {
+            this.serviceProvider = serviceProvider;
+
+            var componentModel = (IComponentModel)serviceProvider.GetService(typeof(SComponentModel));
+            workspace = componentModel.GetService<VisualStudioWorkspace>();
+
+            BuildProjectPathToIdMap();
+        }
+
+        /// <summary>
+        /// Attempts to fetch the extra information required to map from a Roslyn issue
+        /// to a SonarQube server issue. Returns null if there is not enough information
+        /// to produce the mapping.
+        /// </summary>
+        public LiveIssue Create(Diagnostic diagnostic)
+        {
+            // Get the file and project containing the issue
+            SyntaxTree tree = diagnostic.Location?.SourceTree;
+            if (tree == null)
+            {
+                return null;
+            }
+
+            if (diagnostic.Location == Location.None)
+            {
+                return null;
+            }
+
+            Project project = workspace?.CurrentSolution?.GetDocument(tree)?.Project;
+            if (project == null)
+            {
+                return null;
+            }
+
+            string projectId;
+            if (!projectPathToProjectIdMap.TryGetValue(project.FilePath, out projectId))
+            {
+                Debug.Fail($"Expecting to have a project id for the Roslyn project: {project.FilePath}");
+                return null;
+            }
+
+            // Get the whole of the line of text containing the issue (needed to compute the checksum)
+            // TODO: checksum for file level issues?
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this 'TODO' comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)